controller_manager/src/controller_manager.cpp

```diff
@@ -2390,21 +2390,65 @@ controller_interface::return_type ControllerManager::update(
         run_controller_at_cm_rate ? period
                                   : rclcpp::Duration::from_seconds((1.0 / controller_update_rate));
 
-      const rclcpp::Time current_time = get_clock()->now();
+      rclcpp::Time current_time;
       bool first_update_cycle = false;
+      if (get_clock()->started())
+      {
+        current_time = get_clock()->now();
+      }
+      else if (
+        time == rclcpp::Time(0, 0, this->get_node_clock_interface()->get_clock()->get_clock_type()))
+      {
+        throw std::runtime_error(
+          "No clock received, and time argument is zero. Check your controller_manager node's "
+          "clock configuration (use_sim_time parameter) and if a valid clock source is "
+          "available. Also pass a proper time argument to the update method.");
+      }
+      else
+      {
+        // this can happen with use_sim_time=true until the /clock is received
+        rclcpp::Clock clock = rclcpp::Clock();
+        RCLCPP_WARN_THROTTLE(
+          get_logger(), clock, 1000,
+          "No clock received, using time argument instead! Check your node's clock "
+          "configuration (use_sim_time parameter) and if a valid clock source is available");
+        current_time = time;
+      }
       if (
         *loaded_controller.last_update_cycle_time ==
         rclcpp::Time(0, 0, this->get_node_clock_interface()->get_clock()->get_clock_type()))
       {
+        // last_update_cycle_time is zero after activation
         first_update_cycle = true;
-        // it is zero after activation
-        *loaded_controller.last_update_cycle_time = current_time;
+        if (
+          current_time <
+          rclcpp::Time(0, 0, this->get_node_clock_interface()->get_clock()->get_clock_type()) +
+            controller_period)
+        {
+          // can't store negative time
+          *loaded_controller.last_update_cycle_time = current_time;
+        }
+        else
+        {
+          *loaded_controller.last_update_cycle_time = current_time - controller_period;
+        }
         RCLCPP_DEBUG(
           get_logger(), "Setting last_update_cycle_time to %fs for the controller : %s",
           loaded_controller.last_update_cycle_time->seconds(), loaded_controller.info.name.c_str());
       }
-      const auto controller_actual_period =
-        (current_time - *loaded_controller.last_update_cycle_time);
+      rclcpp::Duration controller_actual_period(0, 0);
+      if (current_time > *loaded_controller.last_update_cycle_time)
+      {
+        // the nominal case
+        controller_actual_period = (current_time - *loaded_controller.last_update_cycle_time);
+      }
+      else
+      {
+        controller_actual_period = controller_period;
+        RCLCPP_DEBUG(
+          get_logger(), "Using default controller_period %fs for the controller : %s",
+          controller_actual_period.seconds(), loaded_controller.info.name.c_str());
+      }
```

Comment:
```suggestion
      const auto controller_actual_period =
        (current_time - *loaded_controller.last_update_cycle_time);
```

Reply:
I don't think it is needed anymore as in the previous case, we cannot have a current time less than the last update cycle time.

What do you think?